python/GafferOSLUI/OSLObjectUI.py

```diff
@@ -41,6 +42,111 @@
 
 import GafferOSL
 
+import imath
+import functools
+
+_primitiveVariableNamesOptions = {
+	"P" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Point ),
+	"Pref" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Point ),
+	"velocity" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Vector ),
+	"uv" : IECore.V2fData( imath.V2f(), IECore.GeometricData.Interpretation.UV ),
+	"width" : IECore.FloatData(),
+	 "Cs" : IECore.Color3fData(),
+	 "customInt" : IECore.IntData(),
+	 "customFloat" : IECore.FloatData(),
+	 "customVector" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Vector ),
+	 "customNormal" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Normal ),
+	 "customPoint" : IECore.V3fData( imath.V3f(), IECore.GeometricData.Interpretation.Point ),
+	 "customUV" : IECore.V2fData( imath.V2f(), IECore.GeometricData.Interpretation.UV ),
+	 "customColor" : IECore.Color3fData(),
+	 "customMatrix" : IECore.M44fData(),
+	 "customString" : IECore.StringData(),
+	 "multiInput" : None,
+}
+
+##########################################################################
+# _PrimitiveVariablesFooter
+##########################################################################
+
+class _PrimitiveVariablesFooter( GafferUI.PlugValueWidget ) :
+
+	def __init__( self, plug ) :
+
+		row = GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal )
+
+		GafferUI.PlugValueWidget.__init__( self, row, plug )
+
+		with row :
+
+				GafferUI.Spacer( imath.V2i( GafferUI.PlugWidget.labelWidth(), 1 ) )
+
+				menuButton = GafferUI.MenuButton(
+					image = "plus.png",
+					hasFrame = False,
+					menu = GafferUI.Menu(
+						Gaffer.WeakMethod( self.__menuDefinition ),
+						title = "Add Input"
+					),
+					toolTip = "Add Input"
+				)
+				menuButton.setEnabled( not Gaffer.MetadataAlgo.readOnly( plug ) )
+
+				GafferUI.Spacer( imath.V2i( 1 ), imath.V2i( 999999, 1 ), parenting = { "expand" : True } )
+
+	def _updateFromPlug( self ) :
+
+		self.setEnabled( self._editable() )
+
+	def __menuDefinition( self ) :
+
+		result = IECore.MenuDefinition()
+		usedNames = set()
+		for p in self.getPlug().children():
+			usedNames.add( p["name"].getValue() )
```

Comment:
Thinking about it, we probably shouldn't count a name as already "used" based on a current expression value?  If I reject any plugs with non-null inputs in this loop, then the context shouldn't matter, right?

Reply:
I'm not sure I agree - it doesn't matter where the name came from, if it's there, then it's in use. All an input tells us is that some other names could also be in use in other contexts. If you reject all plugs with non-null inputs, then you'll also be skipping promoted plugs, which might not be driven by an expression at all. [This](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/Switch.cpp#L375-L376) is the reliable way to determine if something might be context varying - I've been thinking for a while it might be worth putting in PlugAlgo.

I guess it may not even be sufficient to scope our context if someone is using "scene:path" to drive the name though. I think that, and the avoidance of synchronous computes, is a stronger argument than anything about the name not being used if it comes from an expression.

Reply:
Conclusion of conversation with John:  To avoid computes, I'll imitate the logic from switch for avoiding context dependend plugs.